### PR TITLE
Feat - Check transaction params on third party block explorers

### LINKED ISSUE #5994: Check incoming transactions on other blockexplorers
To mitigate possible attacks is Insight is compromised,  recheck incoming transactions in TX detail views in different public block explorers, like https://blockr.io/.

If the TX is reported confirmed by Insight, and it does not appear at other block explorer: Show an important warning.

If the TX is reported unconfirmed, just a minor warning.

if the TX is at other block explorer, and match (amount, destination) add a Check ("also seen at xxxxxxx.com").

### PR DESCRIPTION
**Pull request by JDonadio**
fixes #5994

### CODE REVIEW

File: src/js/services/blockExplorerService.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    var request = {
+      url: url,
+      method: 'GET',
+      json: true
+    };
+
+    $http(request).then(function(resp) {
+      $log.debug('Tx params:', resp);
+      return cb(null, resp);
+    }, function(err) {
```
Comment (matiu) [inline src/js/services/blockExplorerService.js]: We should separate the err here:

if (err ... = 404) then return cb(null, { notFound: true})
else return(err).

File: src/js/controllers/tx-details.js [hunk 2]
```diff
@@ -1,20 +1,36 @@
 'use strict';
```
Comment (matiu) [inline src/js/controllers/tx-details.js]: There is way to much code in tx-details. js regarding this feature. it should be only a couple of lines.

File: src/js/controllers/tx-details.js [hunk 3]
```diff
@@ -36,6 +54,69 @@ angular.module('copayApp.controllers').controller('txDetailsController', functio
...
+    $scope.loadingService[name] = state;
+    $timeout(function() {
+      $scope.$apply();
+    });
+  };
+
+  function verifyByThirdParties() {
+    if (lodash.isEmpty($scope.availableServices) || $scope.wallet.network == 'testnet') return;
+
+    lodash.each($scope.availableServices, function(service) {
```
Comment (matiu) [inline src/js/controllers/tx-details.js]: this loop should be inside the service.

File: src/js/controllers/tx-details.js [hunk 4]
```diff
@@ -36,6 +54,69 @@ angular.module('copayApp.controllers').controller('txDetailsController', functio
...
+      setLoading(service.name, true);
+
+      verifyByThirdPartyService.getTx(service, txId, function(err, params) {
```
Comment (matiu) [inline src/js/controllers/tx-details.js]: The service api should be something like:


thirdPartyExplorers.verify(tx, function(err, result) {
   if (err) return ...;

   $scope.verifiedOK = result.ok;
   $scope.verifiedNOK = result.nok;
   $scope.verifiedNA = result.na;
});

File: src/js/controllers/tx-details.js [hunk 5]
```diff
@@ -36,6 +54,69 @@ angular.module('copayApp.controllers').controller('txDetailsController', functio
...
+          checkError();
+        } else
+          compareAndUpdateParams(params);
+
+        serviceCounter += 1;
+      });
+    });
+  };
+
+  function compareAndUpdateParams(params) {
```
Comment (matiu) [inline src/js/controllers/tx-details.js]: inside the service.

Please do not use `params`. Parameters are a function input.  It is a generic name.

File: src/js/services/verifyByThirdPartyService.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  var getTxFromBlockchainInfo = function(txid, cb) {
+    var service = lodash.find(configService.getDefaults().blockExplorerServices, function(s) {
+      return (s.name == 'Blockchain.info');
+    });
+
+    requestTx(service.name, service.url + txid, function(err, resp) {
+      if (err) return cb(err);
+      if (resp.notFound) return cb(null, resp);
+
+      var params = {};
```
Comment (matiu) [inline src/js/services/verifyByThirdPartyService.js]: please dont use `params` name  as a response.

File: src/js/services/verifyByThirdPartyService.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  var explorers = {
+    'Blockchain.info': {
+      getTx: getTxFromBlockchainInfo,
+    },
+    'Blockr.io': {
+      getTx: getTxFromBlockrIo,
+    }
+  };
+
+  function requestTx(serviceName, url, cb) {
```
Comment (matiu) [inline src/js/services/verifyByThirdPartyService.js]: why parameter serviceName here?

File: src/js/services/configService.js [hunk 8]
```diff
@@ -68,6 +68,21 @@ angular.module('copayApp.services').factory('configService', function(storageSer
       enabled: false,
     },
 
+    blockExplorerServices: [
```
Comment (matiu) [inline src/js/services/configService.js]: the service listing  suggest that adding a service is just adding a new item in the config, and it is not.

File: src/js/services/verifyByThirdPartyService.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        params.out.push({
+          address: out.address,
+          amount: parseInt(out.amount * unitToSatoshi)
+        });
+      });
+      return cb(null, params);
+    });
+  };
+
+  var explorers = {
```
Comment (matiu) [inline src/js/services/verifyByThirdPartyService.js]: here there shoudl be all the info about the services:


name,
url,
parser: function(response) {
  return normalizedJason;
}

**Comment by JDonadio** [comment]
Updated. Thanks for your comments!

File: www/views/advancedSettings.html [hunk 10]
```diff
@@ -25,6 +25,15 @@
 
       <div class="item item-divider"></div>
 
+      <ion-toggle class="has-comment" ng-model="verifyTransactionEnabled.value" toggle-class="toggle-balanced" ng-change="verifyTransactionChange()">
+        <span class="toggle-label" translate>Verify transaction</span>
+      </ion-toggle>
+      <div class="comment" translate>
+        If enabled, each transaction will be automatically verified by third party Block Explorer.
```
Comment (matiu) [inline www/views/advancedSettings.html]: Block Explorer*s*

Comment (JDonadio) [inline www/views/advancedSettings.html]: Updated. thanks!

**Pull request closed without merging**